flax/experimental/nnx/nnx/state.py

```diff
@@ -74,29 +74,23 @@ def __init__(
   def raw_mapping(self) -> dict[Key, dict[str, tp.Any] | tp.Any]:
     return self._mapping
 
-  def __getitem__(self, key: Key | int) -> Variable | State:
+  def __getitem__(self, key: Key | int) -> State:
     if isinstance(key, int):
       key = str(key)
     value = self._mapping[key]
-    if isinstance(value, Variable):
-      return value
-    return State(value)
+    if isinstance(value, dict):
+      return State(value)
+    return value
 
-  def __getattr__(self, key: Key) -> Variable | State:
+  def __getattr__(self, key: Key) -> tp.Any:
```

Comment:
Is the return type genuinely `Any` here? It looks it could be `StateLeaf | State` where

```
StateLeaf = Union[Variable, jax.Array, np.ndarray]
```

Reply:
wow, coded this exact solution after implementing the TypeGuard :)

Reply:
...but hadn't seen you suggested exactly this